gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java

```diff
@@ -220,119 +183,39 @@ static GoogleCloudStorageGrpcReadChannel open(
-      Watchdog watchdog,
-      GoogleCloudStorageReadOptions readOptions,
-      BackOffFactory backOffFactory)
-      throws IOException {
-    StorageBlockingStub stub = stubProvider.newBlockingStub();
-    checkArgument(storage != null, "GCS json client cannot be null");
-    checkArgument(itemInfo != null, "object metadata cannot be null");
-    // The non-gRPC read channel has special support for gzip. This channel doesn't
-    // decompress gzip-encoded objects on the fly, so best to fail fast rather than return
-    // gibberish unexpectedly.
-    StorageResourceId resourceId = itemInfo.getResourceId();
-    if (!itemInfo.exists()) {
-      throw new FileNotFoundException(
-          String.format(
-              "%s not found: %s", itemInfo.isDirectory() ? "Directory" : "File", resourceId));
-    }
-    String contentEncoding = itemInfo.getContentEncoding();
-    if (contentEncoding != null && contentEncoding.contains("gzip")) {
-      throw new IOException(
-          "Cannot read GZIP encoded files - content encoding support is disabled.");
-    }
-
+    validate(itemInfo);
+    this.useZeroCopyMarshaller =
+        ZeroCopyReadinessChecker.isReady() && readOptions.isGrpcReadZeroCopyEnabled();
+    this.stub = stubProvider.newBlockingStub();
+    this.resourceId = itemInfo.getResourceId();
+    this.objectGeneration = itemInfo.getContentGeneration();
+    this.objectSize = itemInfo.getSize();
+    this.watchdog = watchdog;
+    this.readOptions = readOptions;
+    this.backOffFactory = backOffFactory;
+    this.readStrategy = readOptions.getFadvise();
     int prefetchSizeInBytes = readOptions.getMinRangeRequestSize() / 2;
-    long objectSize = itemInfo.getSize();
-    long footerOffsetInBytes = max(0, (objectSize - prefetchSizeInBytes));
-
-    long startTime = System.currentTimeMillis();
-    ByteString footerContent = getFooterContent(resourceId, readOptions, stub, footerOffsetInBytes);
-    long endTime = System.currentTimeMillis();
-    if (footerContent == null) {
-      logger.atFiner().log(
-          "Prefetched footer content is null for resource '%s', spent %d milliseconds",
-          resourceId, (endTime - startTime));
-    } else {
-      logger.atFiner().log(
-          "prefetched footer, resource:%s, time:%d, size:%d",
-          resourceId, (endTime - startTime), footerContent.size());
-    }
-
-    return new GoogleCloudStorageGrpcReadChannel(
-        stub,
-        resourceId,
-        itemInfo.getContentGeneration(),
-        itemInfo.getSize(),
-        footerOffsetInBytes,
-        footerContent,
-        watchdog,
-        readOptions,
-        backOffFactory);
+    this.gRPCReadMessageTimeout = readOptions.getGrpcReadMessageTimeoutMillis();
+    this.footerStartOffsetInBytes = max(0, (objectSize - prefetchSizeInBytes));
+    int footerSize = Math.toIntExact(min(prefetchSizeInBytes, objectSize));
+    this.footerBuffer = getFooterContent(footerStartOffsetInBytes, footerSize);
```

Comment:
I think reading the footer (2MB) on `open()` may impact connector performance. It maybe better to pre-fetch footer only when a read is happening in the footer range. 

For example of a full file is being read, the read will start from 0 offset and will continue till the end of the file. In that case pre-fetching footer does not give any advantage. It may infact hamper performance due to switching from a streaming read to fetching data from footerBuffer.

If a parquet file is being read, then the first read operation will happen with an offset in the footer range. In that case we prefetch the footer in that read operation and subsequent reads from footer range can be served from footer buffer.

If some other arbitrary range reads are happening, then again pre-fetching the footer is not going to give us any added advantage. The footer that we pre-fetch may be used only once to read.

Reply:
> For example of a full file is being read, the read will start from 0 offset and will continue till the end of the file. In that case pre-fetching footer does not give any advantage. It may infact hamper performance due to switching from a streaming read to fetching data from footerBuffer.

I don't think there is any perceivable impact here. Footer prefetch in sequential read would not add any advantage but it won't hinder performance either. Total network calls and bytes read would be same.


 > If a parquet file is being read, then the first read operation will happen with an offset in the footer range. In that case we prefetch the footer in that read operation and subsequent reads from footer range can be served from footer buffer.

This is the ideal case, which we typically hit with 1MB default for footer size, and support configuration to manipulate it 

> If some other arbitrary range reads are happening, then again pre-fetching the footer is not going to give us any added advantage. The footer that we pre-fetch may be used only once to read.

What would seem like Arbitrary range reads are offsets determined based on the footer metadata and query from Spark.

For clarification, the approach for footer pre-fetch is not in contention here, if we do want to explore alternatives, we should get back on the original design. The scope of this change is limited to adding support for current footer prefetch with message level timeouts and zero-copy deserializer